compiler/rustc_codegen_llvm/src/debuginfo/metadata/enums/cpp_like.rs

```diff
@@ -784,8 +786,13 @@ fn build_union_fields_for_direct_tag_enum_or_generator<'ll, 'tcx>(
         let field_name = variant_union_field_name(variant_member_info.variant_index);
         let (size, align) = size_and_align_of(enum_type_and_layout);
 
+        let ty::Adt(enum_adt_def, _) = enum_type_and_layout.ty.kind() else {
+            unreachable!();
+        };
```

Comment:
I don't think this will work because this function is shared between enums and generators so `ty.kind()` can return `ty::Generator`. Generators kind of throw a wrench into this approach as their variants don't have `DefIds`. If you want to support that in this PR, we probably need to refactor some of the DI builder functions to take either a `DefId` or just a `(file, line_number)` pair, perhaps as a new `enum DefinitionLocation` type . We can also just leave that for future work 🙂

Reply:
The codegen test isn't covering this properly right now. Interestingly, it panics on Windows, but not on Linux. It seems like I might be able to use `variant_member_info.source_info` (which is of type `Option<(&'ll DIFile, c_uint)>`), since we're iterating members here. The test still passes on Linux when I make the changes necessary to use that, but as you pointed out elsewhere I need to update the codegen test to work properly on Windows. It seems like this function is potentially only properly exercised on Windows for a reason I don't currently understand.

Reply:
The way we encode enums is very different between Windows and Linux (and other *nix likes). Linux uses DWARF which has native support for variant types, which we just call enums in Rust. Windows' debuginfo format PDB is very C++ centric and doesn't have an equivalent concept. As such, we represent Rust enums as unions consisting of one field per variant plus the discriminator field. We then bundle some visualization scripts into the debuginfo that can display the correct value to the user in a debugger. 

This split is evidenced in the file names "cpp_like" is second encoding (the Windows one) and "native" is for platforms that have native support for variant types that LLVM can lower to. As such, any bugs in one of these files will only show up on the corresponding platform.